cfme/modeling/base.py

```diff
@@ -0,0 +1,143 @@
+            raise Exception('Collection [{}] not known to applinace'.format(name))
+
+
+class ObjectCollections(ApplianceCollections):
+    def __init__(self, parent):
+        self._collection_cache = {}
+        self.parent = parent
+        self.appliance = self.parent.appliance
+        self.collections = self.parent._collections
+        self.load_collections()
+
+    def load_collections(self):
+        for collection, cls_and_or_filter in self.collections.items():
+            filter = {'parent': self.parent}
+            if isinstance(cls_and_or_filter, tuple):
+                filter.update(cls_and_or_filter[1])
+                cls = cls_and_or_filter[0]
+            else:
+                cls = cls_and_or_filter
+            collection_instance = cls(self.appliance, filters=filter)
+            self._collection_cache[collection] = collection_instance
+
+
+@attr.s
+class BaseCollection(NavigatableMixin):
+    """Class for helping create consistent Collections
+
+    The BaseCollection class is responsible for ensuring two things:
+
+    1) That the API consistently has the first argument passed to it
+    2) That that first argument is an appliance instance
+
+    This class works in tandem with the entrypoint loader which ensures that the correct
+    argument names have been used.
+    """
+
+    ENTITY = None
+
+    parent = attr.ib()
+    filters = attr.ib(default=attr.Factory(dict))
+
+    @property
+    def appliance(self):
+        if isinstance(self.parent, BaseEntity):
+            return self.parent.appliance
+        else:
+            return self.parent
+
+    @classmethod
+    def for_appliance(cls, appliance, *k, **kw):
+        return cls(appliance)
+
+    @classmethod
+    def for_entity(cls, obj, *k, **kw):
+        return cls(obj, *k, **kw)
+
+    @classmethod
+    def for_entity_with_filter(cls, obj, filt, *k, **kw):
+        return cls.for_entity(obj, *k, **kw).filter(filt)
+
+    def instantiate(self, *args, **kwargs):
+        return self.ENTITY.from_collection(self, *args, **kwargs)
+
+    def filter(self, filter):
+        filters = self.filters.copy()
+        filters.update(filter)
+        return attr.evolve(self, filters=filters)
+
+
+@attr.s
+class BaseEntity(NavigatableMixin):
+    """Class for helping create consistent entitys
+
+    The BaseEntity class is responsible for ensuring two things:
+
+    1) That the API consistently has the first argument passed to it
+    2) That that first argument is a collection instance
+
+    This class works in tandem with the entrypoint loader which ensures that the correct
+    argument names have been used.
+    """
+
+    parent = attr.ib(repr=False)  # This is the collection
+
+    @property
+    def appliance(self):
+        return self.parent.appliance
+
+    @classmethod
+    def from_collection(cls, collection, *k, **kw):
+        return cls(collection, *k, **kw)
+
+    @cached_property
+    def collections(self):
+        return ObjectCollections(self)
```

Comment:
we should extract the classlevel objec exactly here and pass it as ctor arg

Reply:
So I tried adding it as an ```attr.ib()``` and use a Factory to instantiate ObjectsCollections class, but it wasn't appreciated :)